### ISSUE TITLE
Mug frequently unavailable for 20 seconds

### ISSUE THREAD

**Issue by jhemak** [sopelj/hass-ember-mug-component#59]
* Home Assistant Version (eg. 2023.11): 2024.10.4
* Ember Mug Component Version (eg. 0.8.0): 1.0.0
* Which device(s) are you using for Bluetooth (e.g. Bluetooth Dongle / ESPHome Proxy): ESPHome Bluetooth Proxy

My mug and bluetooth proxy are in the same room and yet my mug frequently goes unavailable ... this can happen every few minutes up to a few times per hour.  The mug automatically reconnects and become available ~20 seconds later.

Have you seen this before and do you have any ideas for how to make it stop?

**Comment by sopelj** [comment]
Hi! That's a bit strange. Do you have multiple proxies or dongles close to the mug that might be causing interference? Is this when the mug is in standby, when it has liquid, or both?

**Comment by jhemak** [comment]
I know!  I have only one proxy and I am seeing the behavior even with the mug sitting 6" away from the proxy (though it's typically more like 6' away).  I tried re-flashing the proxy just to see if that changed anything (it did not).  I see the problem when the mug is empty as well as when it has liquid.

**Comment by jhemak** [comment]
Hi @sopelj - I am no expert, but I have a suspicion the issue here is related to the links below:

https://github.com/home-assistant/core/issues/87158#issuecomment-1414519465
https://github.com/home-assistant/core/issues/87158#issuecomment-1414921734
https://github.com/home-assistant/core/issues/87158#issuecomment-1441039007
https://github.com/esphome/bluetooth-proxies/issues/67

My proxy is Wifi, not Ethernet.  If I'm reading these posts right, the device needs to switch between Bluetooth scanning and Wifi.  When not scanning, it may be missing updates from the mug and then causing the integration to think the mug is unavailable.  Does that make sense?  If there is a way you could extend the amount of time before your integration considers the device unavailable that might resolve the problem.

Please let me know what you think.

**Comment by sopelj** [comment]
Oh that's interesting. It's true that WiFi can interfere with Bluetooth in proxies. I do have one of my proxies that I adjusted the strength as per the example in that ticket. Have you tried lowering the strength? What are you using as a proxy? Some devices may have more issues than others. I have quite a few around the house and haven't had any issues with connectivity and they're all WiFi. Do you have other Bluetooth devices that connect to the proxies that could cause problems?

**Comment by jhemak** [comment]
Can you point me to the instructions you are referencing for lowering the signal strength?  It doesn't seem like the issue is so much interference as it is that the device cannot simultaneously process wifi and bluetooth at the same time and it switches back and forth and if it's processing wifi at the time a ble message is sent it will be missed.

I am just using a simple ESP32 board flashed with the stock firmware:
https://esphome.io/projects/?type=bluetooth

Is there a place I could tweak the "unavailability timeout" you are using to see if increasing it would revolve what I'm seeing?

**Comment by sopelj** [comment]
You can adjust it in ESPHome. Just edit your proxies yaml and install it. For example, in one of mine I have put:

```yaml
esp32_ble_tracker:
  scan_parameters:
    interval: 460ms
    window: 160ms
    active: true
    continuous: true
```

I'm not saying those are the ideal settings, but you can play around with them.

There is a request for that in #55, but I have not yet had time to implement it. It's not really an ideal solution as it doesn't solve connectivity issues, it just hides them.

You can also try enabling debug logging and seeing what kind of errors you have. If it's just connection issues or what. If you enable connection state sensors on your proxy you can also see if the proxy is staying connected to home assistant, perhaps it is cutting out too, which may indicate a WiFi issue.

**Comment by jhemak** [comment]
OK - I will give that a shot... although it's here that it recommends not changing these settings from the default when using Wifi proxies.  I understand your point in #55, but that seems a bit different.  In my case, the mug does not have a connection issue per se... it's near the proxy and connected, but as the proxy flips between Wifi and Bluetooth processing it seems to be making HA *think* there is a connection issue when really there is not.

**Comment by jhemak** [comment]
Unfortunately, even with your settings I am seeing the same behavior.  The proxy itself has a strong and solid wifi connection and no errors in HA.  However the mug goes unavailable for ~20 seconds on a regular basis.

**Comment by sopelj** [comment]
Just out of curiosity, where did you find the information that the proxies can only do WiFi or Bluetooth? The only information I could find that the ESP32 could not do WiFi and Bluetooth simultaneously [is here](https://esphome.io/components/esp32_ble_tracker.html#use-on-single-core-chips), and it seems to just apply to single core boards.

Also, it may have issues, but it may be worth trying the latest beta. I'm trying to improve the update mechanism, so it may resolve things, but not if it actually disconnects fully.

**Comment by jhemak** [comment]
It's certainly possible I am wrong ... I don't have any good sources besides the ones I linked above.

I downloaded the latest beta... it definitely seems like an improvement!  It's been 8+ hours since my last `unavailable`, which is a record for sure.  

Unfortunately, it now seems like expected attribute updates (e.g. `Unplugged` to `Plugged in`), which previously were "instant" are now taking 15-30 seconds.

**Comment by jhemak** [comment]
FYI ... just grabbed 1.2.0-beta-3 and the frequent `unavailable` problem is back as it was with the production release.

**Comment by sopelj** [comment]
No, I believe you, i am just trying to identify the issue. Those issues are two years old and closed. It could still be that, but the proxy component has improved a lot in that time. Is your board a single CPU one?

Thanks for testing, I appreciate it. That's good you didn't have unavailable with beta 2, however, the only reason you didn't is because it was receiving events, but no longer updating them. It does mean that it doesn't go unavailable, but it also means you don't get useful updates. Beta 3 actually updates the state again, which is why you're seeing updates again.

**Comment by jhemak** [comment]
No, I am 99% sure I do not have a single CPU board.

**Comment by jhemak** [comment]
Thanks for your work on this.  Beta 3 seems to have fewer `unavailable` disconnects than the production version, though they are not gone completely.  For example, I have had only 6 all day today.  In the production version, I would have seen it that many times, if not more, per hour.

**Comment by sopelj** [comment]
That's a good sign! I appreciate the feedback. Do you know if your Proxy is still disconnecting for long periods? If you want to get stats on it, you can use the [ble_presence](https://esphome.io/components/binary_sensor/ble_presence.html) for the Mug and the [status sensor](https://esphome.io/components/binary_sensor/status.html) in ESPHome. This will help get information on when it is online and when it seems the mug. It might help us pinpoint the issue
